### PR TITLE
Feature/csv contract export

### PR DESCRIPTION
**Pull request by LindaAchilles**
Added CSV Export for contract cost information (to be found under the menu item "export":

image

Added list, detail, csv download and delete views. Contracts can be filtered by date (refers to the invoice dates), funding source and invoice payment status:

### CODE REVIEW

File: tests/exports/contract_csv/test_export_service.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    create_invoice_with_funded_position(contract_year)
+
+    export = export_contract_to_csv(InvoiceSearchParams())
+
+    df = pl.read_csv(StringIO(export), separator=";")
+
+    assert df.height == 3
+
+    funded_amounts = sorted(Decimal(str(v)) for v in df["funded_amount"].to_list())
```
Comment (SvenMarcus) [inline tests/exports/contract_csv/test_export_service.py]: reproduces production code behavior?

File: tests/exports/contract_csv/test_flatteners.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    assert row["invoice_number"] == invoice.number
+    assert row["invoice_date"] == invoice.date.isoformat()
+    assert row["creditor"] == invoice_model.creditor.name
+    assert row["invoice_status"] == invoice.status.value
+    assert row["invoice_currency"] == invoice.currency().code
+    assert row["invoice_comment"] == invoice.comment or ""
+    assert row["external_invoice_id"] == invoice.external_invoice_id or ""
+
+    assert Decimal(row["position_amount"]) == invoice_position.cost.amount
+    assert Decimal(row["tax_rate"]) == invoice_position.tax_rate * 100
```
Comment (SvenMarcus) [inline tests/exports/contract_csv/test_flatteners.py]: use tax_rate.percentage()

File: tests/exports/contract_csv/test_flatteners.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    assert row["start_date"] == str(contract.period.start)
+    assert row["end_date"] == str(contract.period.end)
+    assert row["publishers"] == "Test Publisher"
+    assert row["journals"] == "Test Journal"
+    assert row["publication_billing"] == str(contract.publication_billing)
+    assert row["active_status"] == str(contract.is_active())
+
+    assert row["invoice_number"] == invoice.number
+    assert row["invoice_date"] == invoice.date.isoformat()
+    assert row["creditor"] == invoice_model.creditor.name
```
Comment (SvenMarcus) [inline tests/exports/contract_csv/test_flatteners.py]: fetch creditor by going through invoice instead

File: src/coda/apps/templates/exports/generate_export_form.html [hunk 4]
```diff
@@ -28,17 +28,33 @@ <h2>{{ parameters_title }}</h2>
...
+                {% endif %}
+                {% if "labels" in show_filters or not show_filters %}
+                    {% include "fundingrequests/forms/label_dropdown.html" %}
+                {% endif %}
+                {% if "payment_status" in show_filters or not show_filters %}
+                    {% include payment_status_filter_template|default:"fundingrequests/forms/payment_status.html" %}
+                {% endif %}
+                {% if "funding_source" in show_filters or not show_filters %}
+                    {% include "invoices/partials/funding_source_filter.html" %}
+                {% endif %}
```
Comment (SvenMarcus) [inline src/coda/apps/templates/exports/generate_export_form.html]: simplify with for loop

File: tests/exports/contract_csv/test_mappers.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    assert contract_dto.start_date == contract.period.start
+    assert contract_dto.end_date == contract.period.end
+    assert contract_dto.publishers == ["Test Publisher"]
+    assert contract_dto.journals == ["Test Journal"]
+    assert contract_dto.publication_billing == contract.publication_billing
+    assert contract_dto.active == contract.is_active()
+
+
+@pytest.mark.django_db
+def test__contract_with_esac__maps_to_dto__all_required_fields_are_mapped_correctly() -> None:
```
Comment (SvenMarcus) [inline tests/exports/contract_csv/test_mappers.py]: parametrize identical tests by link type

File: tests/exports/contract_csv/test_mappers.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+from coda.domain.publication.publication import JournalId
+from tests import domainfactory, modelfactory
+from coda.apps.contracts import repository as contract_repository
+from coda.apps.contracts.models import Contract as ContractModel, ContractLink, ContractLinkType
+from coda.apps.invoices.models import Invoice as InvoiceModel
+from tests.exports.helpers import create_invoice_with_contract_position
+
+
+@pytest.mark.django_db
+def test__contract_without_invoices__maps_to_dto__all_required_fields_are_mapped_correctly() -> (
```
Comment (SvenMarcus) [inline tests/exports/contract_csv/test_mappers.py]: better test names?

File: src/coda/apps/exports/services/contract_csv/dtos.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+from pydantic import BaseModel
+from coda.contexts.finance.dto.import_dtos import InvoiceImportDto
+
+
+class ContractLinkDto(BaseModel):
+    type: str
+    value: str
+
+
+class ContractInfoDto(BaseModel):
```
Comment (SvenMarcus) [inline src/coda/apps/exports/services/contract_csv/dtos.py]: Clearer name to highlight usecase. (ContractExportRow?)

File: src/coda/apps/exports/views/contract_csv_views.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    entity_filter_template = "entity_generic_filter.html"
+
+
+contract_csv_export_list_view = ContractCSVExportListView.as_view()
+
+
+@login_required
+@require_GET
+@breadcrumb("CSV Export Details", parent_url_name=CONTRACTS_CSV_LIST_URL)
+def contract_csv_detail_page(request: HttpRequest, pk: int) -> HttpResponse:
```
Comment (SvenMarcus) [inline src/coda/apps/exports/views/contract_csv_views.py:63]: pick either singular or plural naming

**Review by SvenMarcus — CHANGES_REQUESTED** [review]
- Missing links in CSV export
- missing date range boundary test

Missing edge cases: test_queries.py
No date range boundary tests (before/on/after); no "no invoices → empty output" test

File: src/coda/apps/exports/services/contract_csv/flatteners.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+from coda.apps.exports.services.contract_csv.dtos import ContractCSVExportDto, ContractDetailsDto
+from coda.contexts.finance.dto.import_dtos import (
+    ContractPositionImportDto,
+    FundingAssignmentImportDto,
+    InvoiceImportDto,
+)
+
+
+def flatten_contract_data(dto: ContractCSVExportDto) -> list[dict[str, str]]:
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/flatteners.py:11]: maybe as a method of `ContractCSVExportDto`?

File: src/coda/apps/exports/services/contract_csv/mappers.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    contract: Contract, matching_invoice_ids: set[InvoiceId] | None = None
+) -> ContractCSVExportDto:
+    contract_dto = map_contract_to_dto(contract)
+
+    invoices = get_relevant_invoices_for_contract(contract, matching_invoice_ids)
+
+    invoice_dtos: list[InvoiceImportDto] = []
+    for invoice in invoices:
+        scoped_positions = [
+            pos for pos in contract.position_set.all() if pos.invoice_id == invoice.pk
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/mappers.py:23]: Query?

File: src/coda/apps/exports/services/contract_csv/mappers.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+            ContractLinkDto(type=link.type.name, value=link.value)
+            for link in contract_model.links.all()
+        ],
+    )
+
+
+def get_relevant_invoices_for_contract(
+    contract: Contract, matching_invoice_ids: set[InvoiceId] | None = None
+) -> list[Invoice]:
+    invoices = {pos.invoice for pos in contract.position_set.all() if pos.invoice is not None}
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/mappers.py:55]: Query filter

File: src/coda/apps/exports/services/contract_csv/mappers.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    if matching_invoice_ids is not None:
+        invoices = {inv for inv in invoices if inv.pk in matching_invoice_ids}
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/mappers.py:57]: Can be also in the query filter

File: src/coda/apps/exports/views/base_csv_views.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    applied_filters_builder: Callable[[dict[str, str]], list[AppliedFilter]],
+    create_url_name: str,
+) -> HttpResponse:
+
+    export = get_object_or_404(model, pk=pk)
+    export_ = cast(_CSVExportInstance, export)
+
+    preview_df = _create_preview_dataframe(
+        export_.csv_file.open("rb").read().decode("utf-8"), preview_columns
+    )
```
Comment (fjen) [inline src/coda/apps/exports/views/base_csv_views.py:46]: use `with` for file handle and pass the file directly.
```suggestion
    with export_.csv_file.open("rb") as csv_file:
        preview_df = _create_preview_dataframe(csv_file, preview_columns)
```

File: src/coda/apps/exports/views/base_csv_views.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    export_.csv_file.save(
+        filename,
+        ContentFile(csv_content.encode("utf-8")),
+    )
+
+    return redirect(detail_url_name, pk=export.pk)
+
+
+def _create_preview_dataframe(csv_content: str, preview_columns: list[str]) -> pl.DataFrame:
+    return pl.read_csv(StringIO(csv_content), separator=";").select(preview_columns).head(50)
```
Comment (fjen) [inline src/coda/apps/exports/views/base_csv_views.py:125]: Use polars' `n_rows` instead of `head()` to not parse the whole file.

```suggestion
def _create_preview_dataframe(csv_file: BinaryIO, preview_columns: list[str]) -> pl.DataFrame:
    return pl.read_csv(csv_file, separator=";", n_rows=50).select(preview_columns)
```

File: src/coda/apps/exports/services/contract_csv/queries.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+            "publishers",
+            "journals",
+            "links",
+            "links__type",
+            Prefetch(
+                "position_set__invoice",
+                queryset=InvoiceModel.objects.select_related("creditor").prefetch_related(
+                    "currency_conversions",
+                    "positions__funding_assignments__funding_source",
+                    "positions__contract",
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/queries.py:40]: prefetch not used later. mapper uses `scoped_positions` in `map_contract_to_export_dto` which is always non-empty. `invoice.positions.all()` is never evaluated.

File: src/coda/apps/exports/services/contract_csv/queries.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+            ),
+            Prefetch(
+                "position_set__funding_assignments",
+                queryset=FundingAssignment.objects.select_related("funding_source"),
+            ),
+        )
+        .distinct()
+    )
+
+    return contracts, matching_invoice_ids
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/queries.py:51]:
```suggestion
def get_contracts_for_export(params: invoice_query.InvoiceSearchParams) -> QuerySet[Contract]:
    criteria = invoice_query.build_criteria(params)
    matching_invoices = invoice_query.search(*criteria)

    has_matching_position = Exists(
        PositionModel.objects.filter(
            contract_id=OuterRef("pk"),
            invoice__in=matching_invoices,
        )
    )

    contracts = (
        Contract.objects.filter(has_matching_position)
        .prefetch_related(
            "publishers",
            "journals",
            "links",
            "links__type",
            # Scope positions to matching invoices in SQL. This must come
            # before the other position_set lookups: Django caches the first
            # position_set fetch per contract and silently skips later ones.
            Prefetch(
                "position_set",
                queryset=PositionModel.objects.filter(invoice__in=matching_invoices)
                .select_related("invoice__creditor")
                .prefetch_related("invoice__currency_conversions"),
            ),
            Prefetch(
                "position_set__funding_assignments",
                queryset=FundingAssignment.objects.select_related("funding_source"),
            ),
        )
        .distinct()
    )

    return contracts
```

File: src/coda/apps/exports/services/contract_csv/mappers.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
+from coda.apps.contracts.models import Contract
+from coda.apps.exports.services.contract_csv.dtos import (
+    ContractCSVExportDto,
+    ContractDetailsDto,
+    ContractLinkDto,
+)
+from coda.apps.exports.services.fundingrequest_csv.mappers import map_invoice_to_dto
+from coda.apps.invoices.models import Invoice
+from coda.contexts.finance.dto.import_dtos import InvoiceImportDto
+from coda.domain.finance.invoice import InvoiceId
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/mappers.py:10]:
```suggestion
```

File: src/coda/apps/exports/services/contract_csv/mappers.py [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+
+def map_contract_to_export_dto(
+    contract: Contract, matching_invoice_ids: set[InvoiceId] | None = None
+) -> ContractCSVExportDto:
+    contract_dto = map_contract_to_dto(contract)
+
+    invoices = get_relevant_invoices_for_contract(contract, matching_invoice_ids)
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/mappers.py:18]:
```suggestion
def map_contract_to_export_dto(contract: Contract) -> ContractCSVExportDto:
    contract_dto = map_contract_to_dto(contract)

    invoices = get_relevant_invoices_for_contract(contract)
```

File: src/coda/apps/exports/services/contract_csv/mappers.py (hunk 12 above)
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/mappers.py:57]:
```suggestion
def get_relevant_invoices_for_contract(contract: Contract) -> list[Invoice]:
    invoices = {pos.invoice for pos in contract.position_set.all() if pos.invoice is not None}
```

File: src/coda/apps/exports/services/contract_csv/export_service.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    "funded_amount",
+    "funding_source_name",
+    "funding_source_type",
+]
+
+
+def export_contract_to_csv(
+    params: InvoiceSearchParams,
+) -> str:
+    contracts, matching_invoice_ids = queries.get_contracts_for_export(params)
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/export_service.py:37]:
```suggestion
    contracts = queries.get_contracts_for_export(params)
```

File: src/coda/apps/exports/services/contract_csv/export_service.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+    export_dtos = [
+        map_contract_to_export_dto(contract, matching_invoice_ids=matching_invoice_ids)
```
Comment (fjen) [inline src/coda/apps/exports/services/contract_csv/export_service.py:40]:
```suggestion
        map_contract_to_export_dto(contract)
```

**Review by fjen — CHANGES_REQUESTED** [review]
Unfiltered export loads every invoice PK in the database into a set to do python-side membership test in `get_relevant_invoices_for_contract`. 

Do this filtering in the db directly and save passing filtered ids around. Then `contract.position_set.all()` contains only positions on matching invoices, the `matching_invoice_ids` parameter can be dropped from `map_contract_to_export_dto` and `get_relevant_invoices_for_contract`. Result: Query count: 12 -> 7 per export.

tests need to be adjusted to not expect second return value for `get_contracts_for_export(params)`.

File: src/coda/apps/exports/views/fundingrequest_csv_views.py [hunk 21]
```diff
@@ -53,50 +60,25 @@ class FundingRequestCSVExportListView(
...
-    request: HttpRequest,
-) -> HttpResponse:
+@breadcrumb("Generate New CSV Export", parent_url_name=FUNDINGREQUESTS_CSV_LIST_URL)
+def fundingrequest_csv_export_create_view(request: HttpRequest) -> HttpResponse:
 
     if request.method == "GET":
-        context = _get_export_form_context()
+        context = build_filter_form_context()
         context["expand_advanced_search"] = bool(request.GET)
         context["current_filters"] = parse_current_filters_to_context(request)
```
Comment (fjen) [inline src/coda/apps/exports/views/fundingrequest_csv_views.py:83]: is `current_filters` used anywhere?

**Comment by fjen** [comment]
next time please do not force push after initial review. changes since review are not easily visible then. branch cleanup can be done before merge.

File: src/coda/apps/invoices/invoice_query.py [hunk 22]
```diff
@@ -203,6 +203,7 @@ class InvoiceSearchParams:
     contract_year: str | int | None = None
     contract_year_positions_only: bool = False
     has_external_id: bool | None = None
+    decimal_separator: str = "."
```
Comment (fjen) [inline src/coda/apps/invoices/invoice_query.py]: we have a `DecimalSeparator` Enum

File: src/coda/apps/exports/views/base_csv_views.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+    name: str
+    filters: dict[str, str]
+    csv_file: FieldFile
+    record_count: int
+    pk: int
+
+    def delete(self) -> None: ...
+
+
+def csv_detail_page(
```
Comment (fjen) [inline src/coda/apps/exports/views/base_csv_views.py]: all methods in this `base` module are only used by the contract export. can we generalize directly?

**Comment by fjen** [comment]
`contract_csv` directory is missing `__init__.py`

File: src/coda/apps/exports/services/filter_display.py [hunk 24]
```diff
@@ -164,6 +231,37 @@ def build_applied_filters(filters: dict[str, Any]) -> list[AppliedFilter]:
     return applied
 
 
+def build_applied_filters_for_contract(filters: dict[str, Any]) -> list[AppliedFilter]:
```
Comment (fjen) [inline src/coda/apps/exports/services/filter_display.py]: looks like a copy of `build_applied_filters()`. cant we re-use it by extending `_FORMATTERS`?

File: src/coda/apps/exports/services/contract_csv/flatteners.py [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def _create_row(
+    contract: ContractDetailsDto,
+    invoice: InvoiceImportDto,
+    position: ContractPositionImportDto,
+    assignment: FundingAssignmentImportDto | None = None,
+    funded_amount: Decimal | None = None,
+) -> dict[str, str]:
+    link_columns = {link.type.lower(): link.value for link in contract.links}
```
Comment (SvenMarcus) [inline src/coda/apps/exports/services/contract_csv/flatteners.py]: this could lead to duplicate column names if a contract has multiple links of the same type

**Comment by SvenMarcus** [comment]
I will pick this up while @LindaAchilles is on vacation to become a wizarding mermaid rockstar
